File: day-4/puzzle.go
package main

import (
	"flag"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func ParseCard(line string) (winning, ownedNumbers []int) {
	// Trim until ':' symbol reached
	numbers := strings.TrimLeftFunc(line, func(r rune) bool { return r != ':' })
	numbers = numbers[1:] // trim ':'

	sides := strings.Split(numbers, " | ")

	for _, numS := range strings.Split(sides[0], " ") {
		if len(numS) == 0 {
			continue
		}
		num, err := strconv.Atoi(numS)
		if err != nil {
			panic(err)
		}

		winning = append(winning, num)
	}

	for _, numS := range strings.Split(sides[1], " ") {
		if len(numS) == 0 {
			continue
		}
		num, err := strconv.Atoi(numS)
		if err != nil {
			panic(err)
		}

		ownedNumbers = append(ownedNumbers, num)
	}

	return
}

func Intersection(a, b []int) []int {
	// sort a and b
	slices.Sort(a)
	slices.Sort(b)

	result := []int{}

	pa := 0
	pb := 0
	for pa < len(a) && pb < len(b) {
		switch {
		case a[pa] < b[pb]:
			pa++
		case a[pa] > b[pb]:
			pb++
		default:
			// They are equal
			result = append(result, a[pa])
			pa++
			pb++
		}
	}
	return result
}

func main() {
	flag.Parse()
	inputB, err := os.ReadFile(flag.Arg(0))
	if err != nil {
		panic(err)
	}
	input := strings.TrimSpace(string(inputB))
	lines := strings.Split(input, "\n")

	instances := make([]int, len(lines))

	totalPoints := 0
	for cardId, line := range lines {
		instances[cardId]++ // Count the original card
		winning, owned := ParseCard(line)
		intersection := Intersection(winning, owned)

		if len(intersection) == 0 {
			fmt.Printf("Card %d has no winning numbers, so it is worth no points.\n", cardId+1)
			continue
		}

		points := 1
		if len(intersection) > 1 {
			points = 2 << (len(intersection) - 2)
		}
		fmt.Printf("Card %d has winning numbers %v, so it is worth %d point.\n", cardId+1, intersection, points)
		totalPoints += points
		for x := cardId + 1; x <= cardId+len(intersection); x++ {
			fmt.Printf("Card %d has won a card %d\n", cardId+1, x+1)
			instances[x] += instances[cardId]
		}
	}
	fmt.Printf("Total points won: %d\n", totalPoints)
	instancesSum := 0
	for cardId, count := range instances {
		fmt.Printf("There were %d instances of card %d\n", count, cardId+1)
		instancesSum += count
	}
	fmt.Printf("Sum of all cards instances: %d\n", instancesSum)
}

File: day-4/puzzle.lua
---@param a number[]
---@param b integer[]
---@return integer[]
local function intersect(a, b)
	local common = {} ---@type integer[]
	table.sort(a)
	table.sort(b)

	local ap, bp = 1, 1
	while ap <= #a and bp <= #b do
		if a[ap] < b[bp] then
			ap = ap + 1
		elseif a[ap] > b[bp] then
			bp = bp + 1
		else
			-- common value
			table.insert(common, a[ap])
			ap = ap + 1
			bp = bp + 1
		end
	end

	return common
end

---@param line string
---@return integer[]
---@return integer[]
local function parseCard(line)
	local start = assert(line:find(": ", 1, true))
	line = line:sub(start + #": ")

	local sep = assert(line:find(" | ", 1, true))
	local winningS = line:sub(1, sep - 1)
	local ownedS = line:sub(sep + #" | ")

	-- parse winning numbers
	local winning = {} ---@type integer[]
	for num in winningS:gmatch("%d+") do
		table.insert(winning, tonumber(num))
	end

	-- parse owned numbers
	local owned = {} ---@type integer[]
	for num in ownedS:gmatch("%d+") do
		table.insert(owned, tonumber(num))
	end

	return winning, owned
end

local function main()
	if #arg >= 1 then
		local file = assert(io.open(arg[1], "r"))
		io.input(file)
	end

	local metaInstances = {}
	function metaInstances:__index(_)
		return 0
	end
	local instances = {}
	setmetatable(instances, metaInstances)

	local totalPoints, totalScratchcards = 0, 0
	local cardId = 0
	for line in io.lines() do
		cardId = cardId + 1

		print(line)

		instances[cardId] = instances[cardId] + 1
		print("You have " .. instances[cardId] .. " instances of this card")
		totalScratchcards = totalScratchcards + instances[cardId]

		local winning, owned = parseCard(line)

		local common = intersect(winning, owned)
		print("common", table.concat(common, " "))

		if #common == 0 then
			goto continue
		end

		local points = math.floor(2 ^ (#common - 1))
		totalPoints = totalPoints + points
		print("Points", points)

		for extraCardId = cardId + 1, cardId + #common do
			print("Won another instance of card " .. extraCardId)
			instances[extraCardId] = instances[extraCardId] + instances[cardId]
		end

		::continue::
	end
	print("Total points: " .. totalPoints)
	print("Total scratchcards won: " .. totalScratchcards)
end
main()
